linera-core/src/client/mod.rs

```diff
@@ -825,149 +831,164 @@ impl<Env: Environment> Client<Env> {
         Ok(())
     }
 
-    /// Downloads a limited batch of received certificates from a validator.
-    /// Returns the updated tracker and whether there are more certificates to fetch.
-    #[instrument(level = "trace", skip(self))]
-    async fn synchronize_received_certificates_batch_from_validator(
+    /// Downloads and processes certificates for sender chain blocks.
+    #[instrument(level = "trace", skip_all)]
+    async fn download_and_process_sender_chain(
         &self,
-        chain_id: ChainId,
-        remote_node: &RemoteNode<Env::ValidatorNode>,
-    ) -> Result<(ReceivedCertificatesFromValidator, bool), ChainClientError> {
-        let mut tracker = self
-            .local_node
-            .chain_state_view(chain_id)
-            .await?
-            .received_certificate_trackers
-            .get()
-            .get(&remote_node.public_key)
-            .copied()
-            .unwrap_or(0);
-        let (max_epoch, committees) = self.admin_committees().await?;
-
-        // Retrieve a limited batch of received certificates from this validator.
-        let offset = tracker;
-        let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_n(offset);
-        let info = remote_node.handle_chain_info_query(query).await?;
-        let received_entries = info.requested_received_log.len();
-        let has_more = received_entries >= CHAIN_INFO_MAX_RECEIVED_LOG_ENTRIES;
-
-        let remote_log = info.requested_received_log;
-        let remote_heights = Self::heights_per_chain(&remote_log);
+        sender_chain_id: ChainId,
+        nodes: &[RemoteNode<Env::ValidatorNode>],
+        received_log: &ReceivedLogs,
+        mut remote_heights: Vec<BlockHeight>,
+        sender: mpsc::UnboundedSender<ChainAndHeight>,
+    ) {
+        // Verify the certificate before doing any expensive networking.
+        let (max_epoch, committees) = match self.admin_committees().await {
+            Ok(result) => result,
+            Err(error) => {
+                error!(%error, %sender_chain_id, "could not read admin committees");
+                return;
+            }
+        };
+        let committees_ref = &committees;
+        let mut nodes = nodes.to_vec();
+        while !remote_heights.is_empty() {
+            let remote_heights_ref = &remote_heights;
+            nodes.shuffle(&mut rand::thread_rng());
+            let certificates = match communicate_in_parallel(
+                &nodes,
+                async move |remote_node| {
+                    let mut remote_heights = remote_heights_ref.clone();
+                    // No need trying to download certificates the validator didn't have in their
+                    // log - we'll retry downloading the remaining ones next time we loop.
+                    remote_heights.retain(|height| {
+                        received_log.validator_has_block(
```

Comment:
(It still feels like something could be simplified: First we got them from the received log, and now we filter them using the received log. Not a blocker, but maybe something to look into later.)